Update the CQ doc

rkuchan opened the pull request:
Changes include:

* Restructuring the discussion into `Basic syntax` and `Advanced syntax`
* Creating examples that cover common use cases and are more detailed

The diffs won't be very useful - this was a pretty substantial rewrite.

Aims to fix #677, #303, #568, and #611.

rkuchan commented:
@beckettsean, ready for review!

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
```diff
@@ -3,294 +3,786 @@ title: Continuous Queries
+every hour.
+
+When the CQ executes, it runs a single query for the time range between
+[`now()`](/influxdb/v1.0/concepts/glossary/#now) and `now()` minus the
```
This ends up grouping strangely in my head when I read it. I got, "the time range between now and now", minus the group by interval.

I think this might work better:

When the CQ executes, it runs a single query for the time range between [`now()`](/influxdb/v1.0/concepts/glossary/#now) and `now() minus the GROUP BY time() interval`.

beckettsean replied:
Nevermind. I see it throughout the doc, and it was just the line break behavior that made me scan it wrong the first time.

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
-```sql
-CREATE CONTINUOUS QUERY <cq_name> ON <database_name> [RESAMPLE [EVERY <interval>] [FOR <interval>]] BEGIN SELECT <function>(<stuff>)[,<function>(<stuff>)] INTO <different_measurement> FROM <current_measurement> [WHERE <stuff>] GROUP BY time(<interval>)[,<stuff>] END
+```
+CREATE CONTINUOUS QUERY <cq_name> ON <database_name> BEGIN
````
I've been preferring the following syntax recently:

```
CREATE CONTINUOUS QUERY <cq_name> ON <database_name> 
BEGIN
  <cq_query>
END
```

I feel like putting the BEGIN with the END really helps

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
-1. The `INTO` clause:
-This is where you specify the destination measurement for the query results.
+```
+CREATE CONTINUOUS QUERY "cq_basic" ON "transportation" BEGIN
````
BEGIN to next line

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
 
-* Create a CQ with one function and write the results to another [retention policy](/influxdb/v1.0/concepts/glossary/#retention-policy-rp):
+```
+CREATE CONTINUOUS QUERY "cq_basic_rp" ON "transportation" BEGIN
````
BEGIN to next line

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
-> CREATE CONTINUOUS QUERY "minnie_maximus_1" ON "world" BEGIN SELECT min("mouse") AS "minuscule",max("imus") AS "monstrous" INTO "min_max_mouse" FROM "zoo" GROUP BY time(30m) END
-    ```
+```
+CREATE CONTINUOUS QUERY "cq_basic_br" ON "transportation" BEGIN
````
BEGIN to next line

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
-    `vampires_2` runs every 15 minutes and computes two queries per run:
-    one where time is between `now()` and `now() - 30m` and one where time is between `now() - 30m` and `now() - 60m`
+```
+CREATE CONTINUOUS QUERY "cq_basic_offset" ON "transportation" BEGIN
````
BEGIN to next line

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
+### Advanced Syntax
+
+```
+CREATE CONTINUOUS QUERY <cq_name> ON <database_name>
````
BEGIN to next line

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
-```sql
-DROP CONTINUOUS QUERY <cq_name> ON <database_name>
 ```
+CREATE CONTINUOUS QUERY "cq_advanced_every" ON "transportation"
````
BEGIN to next line

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
-While InfluxQL resembles SQL in functionality it does not always have similar
-syntax.
+```
+CREATE CONTINUOUS QUERY "cq_advanced_for" ON "transportation"
````
BEGIN to next line

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
+the CQ's execution interval and the length of the CQ's time range.
+
+```
+CREATE CONTINUOUS QUERY "cq_advanced_every_for" ON "transportation"
````
BEGIN to next line

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
-and writes those averages to the `mean_bees` field in the `aggregate_bees` measurement.
 ```
-CREATE CONTINUOUS QUERY "bee_cq" ON "mydb" BEGIN SELECT mean("bees") AS "mean_bees" INTO "aggregate_bees" FROM "farm" GROUP BY time(30m) END
+CREATE CONTINUOUS QUERY "cq_advanced_for_fill" ON "transportation"
````
BEGIN to next line

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
```diff
@@ -3,294 +3,786 @@ title: Continuous Queries
+##### Issue 3: Backfilling results for older data
+<br>
+CQs operate on realtime data, that is, data with timestamps that occur
+after the time at which you create the CQ.
```
This isn't strictly true. The new CQ will run on data written before it was created, as long as it falls in the relative range sampled. That said, I have no better wording for this section.

rkuchan replied:
Changed to: data with timestamps that occur relative to [`now()`](link-to-glossary-entry).

beckettsean replied:
works for me

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
```diff
@@ -3,294 +3,786 @@ title: Continuous Queries
+
+CQs execute at the same interval as the `EVERY` interval in the `RESAMPLE`
+clause, and they run at the start of InfluxDB’s preset time boundaries.
+If the `EVERY` interval is two hours, InfluxDB executes the CQ at the start of
```
maybe

"InfluxDB executes the CQ at the top of every other hour."

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
```diff
@@ -3,294 +3,786 @@ title: Continuous Queries
+
+Both the `EVERY` interval and the `FOR` interval accept
+[duration literals](/influxdb/v1.0/query_language/spec/#durations).
+The `RESAMPLE` clause works with only the `EVERY` interval, only the `FOR`
```
Maybe

The `RESAMPLE` clause works with either or both of the `EVERY` and `FOR` intervals configured.

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
-DROP CONTINUOUS QUERY <cq_name> ON <database_name>
 ```
+CREATE CONTINUOUS QUERY "cq_advanced_every" ON "transportation"
+RESAMPLE EVERY 1h BEGIN
````
This actually won't work. The `EVERY` setting must be `>=` the `GROUP BY time()` interval. Otherwise the query throws a misleading error. https://github.com/influxdata/influxdb/issues/6963

beckettsean replied:
The example should be something more like executing every minute so that the previous hour's data is still graphable even though the hour hasn't ended.

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
-syntax.
+```
+CREATE CONTINUOUS QUERY "cq_advanced_for" ON "transportation"
+RESAMPLE FOR 1h BEGIN
````
Same issue with `GROUP BY time()` interval smaller than `FOR`

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
+
+```
+CREATE CONTINUOUS QUERY "cq_advanced_every_for" ON "transportation"
+RESAMPLE EVERY 1h FOR 90m BEGIN
````
Same issue with `GROUP BY time()` interval smaller than `FOR`

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
 ```
-CREATE CONTINUOUS QUERY "bee_cq" ON "mydb" BEGIN SELECT mean("bees") AS "mean_bees" INTO "aggregate_bees" FROM "farm" GROUP BY time(30m) END
+CREATE CONTINUOUS QUERY "cq_advanced_for_fill" ON "transportation"
+RESAMPLE FOR 2h BEGIN
````
Same issue with `GROUP BY time()` interval smaller than `FOR`

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
```diff
@@ -3,294 +3,786 @@ title: Continuous Queries
+See [Frequently Encountered Issues](/influxdb/v1.0/troubleshooting/frequently_encountered_issues/#getting-empty-results-with-fill-previous)
+for more information.
+
+## CQ Management
```
Spell it out

Continuous Query Management

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -3,294 +3,786 @@ title: Continuous Queries
+> DROP CONTINUOUS QUERY "idle_hands" ON "telegraf"`
+>
+```
+
````
### Alter CQs

Continuous queries cannot be altered once created. To change a CQ, you must `DROP` and re`CREATE` it with the updated settings.

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
```diff
@@ -323,5 +815,65 @@ Query the data in the measurement `aggregate_bees` and request values of the `me
+the inner-most function and querying the CQ results to calculate the outer-most
+function.
+
+> **InfluxQL functions that support nesting:**
```
_Feels like this should be in the FAQ and there's a link to it from this part of the docs. _

Some InfluxQL functions [support nesting](FAQ) of other functions. 
Most do not. 
If your function does not support nesting, you can get the same functionality using a CQ to calculate
the inner-most function. Then simply query the CQ results to calculate the outer-most
function.

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
````diff
@@ -323,5 +815,65 @@ Query the data in the measurement `aggregate_bees` and request values of the `me
+The query calculates the average number
+of `bees` at `30` minute intervals and the sum of those averages:
+```
+SELECT sum(mean("bees")) FROM "farm" GROUP BY time(30m)
````
to me this query doesn't make sense. There's only one value per 30m bucket for the MEAN() function, so what is SUM() adding?

rkuchan commented:
@beckettsean, I still need to review the advanced examples.

rkuchan commented:
The last commit is pretty substantial and could use some review.

After @beckettsean's comments and after working with my own CQs I realized a couple things:

* If the `EVERY` interval > `GROUP BY time()` interval the CQ runs at the `EVERY` interval and covers the time range between `now()` and `now()` minus the `EVERY` interval, not `now()` minus the `GROUP BY time()` interval. I considered that unexpected behavior and added it to a new Common Issues with Advanced Syntax section.

* If the `EVERY` interval < `GROUP BY time()` interval, the CQ runs at the `EVERY` interval and covers the current time bucket, that is, the time bucket that intersects with `now()`. I considered this expected behavior and changed example one in the advanced syntax as requested.

* If the `FOR` interval > `GROUP BY time()` interval (or the `EVERY` interval if it's there), then everything behaves as expected. That is, the CQ executes at the same interval as the `GROUP BY time()` interval and it covers the time range between `now()` and `now()` minus the `FOR` interval.

* If the `FOR` interval < `GROUP BY time()` interval (or the `EVERY` interval if it's there), the CQ doesn't get created. This is documented in the new Common Issue with Advanced Syntax section and it links to the GitHub issue.

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
```diff
@@ -445,56 +448,69 @@ interval.
-`cq_advanced_every` executes at the start of every hour and calculates the
-average for the previous half hour.
-By design, every other half hour interval will be missing from the CQ results.
+Notice that `cq_advanced_every` calculates the result for the 8:00 time interval
```
This is awesomely explanatory, nice!

beckettsean commented on content/influxdb/v1.0/query_language/continuous_queries.md:
```diff
@@ -707,6 +723,41 @@ previous value is outside the query’s time range.
 
+#### Common Issues with Advanced Syntax
+
+##### Issue 1: If the `EVERY` interval is greater than the `GROUP BY time()` interval
```
Wow, that's crazy. I didn't know that would work. Nice find!

```
> create continuous query foo on mydb resample every 1m begin select count(value) as count_value into foobar from foobar group by time(10s) end 
> insert foobar value=12
> insert foobar value=13
> insert foobar value=14
> insert foobar value=15
> insert foobar value=1123
> insert foobar value=123897
> insert foobar value=1238
> insert foobar value=12384884
> select * from foobar
name: foobar
------------
time				count_value	value
2016-10-03T22:32:50Z		1		
2016-10-03T22:32:58.816013683Z			12
2016-10-03T22:33:06.525668153Z			13
2016-10-03T22:33:07.911455487Z			14
2016-10-03T22:33:09.064298625Z			15
2016-10-03T22:33:10.341394176Z			1123
2016-10-03T22:33:32.35745048Z			123897
2016-10-03T22:33:40.837554187Z			1238
2016-10-03T22:33:46.825528639Z			1.2384884e+07

> select * from foobar
name: foobar
------------
time				count_value	value
2016-10-03T22:32:50Z		1		
2016-10-03T22:32:58.816013683Z			12
2016-10-03T22:33:00Z		3		
2016-10-03T22:33:06.525668153Z			13
2016-10-03T22:33:07.911455487Z			14
2016-10-03T22:33:09.064298625Z			15
2016-10-03T22:33:10Z		1		
2016-10-03T22:33:10.341394176Z			1123
2016-10-03T22:33:30Z		1		
2016-10-03T22:33:32.35745048Z			123897
2016-10-03T22:33:40Z		2		
2016-10-03T22:33:40.837554187Z			1238
2016-10-03T22:33:46.825528639Z			1.2384884e+07

> 
```

beckettsean commented:
Amazing! Ship it!

rkuchan commented:
Thanks for the feedback, @beckettsean!
